server/src/main/java/org/opensearch/action/pagination/WlmPaginationStrategy.java

```diff
@@ -0,0 +1,193 @@
+        if (sortedStats.isEmpty()) {
+            this.responseToken = null;
+            return Collections.emptyList();
+        }
+
+        List<WlmStats> allQueryGroups = new ArrayList<>();
+        for (WlmStats stat : sortedStats) {
+            Map<String, QueryGroupStats.QueryGroupStatsHolder> queryGroups = stat.getQueryGroupStats().getStats();
+            for (Map.Entry<String, QueryGroupStats.QueryGroupStatsHolder> entry : queryGroups.entrySet()) {
+                String queryGroupId = entry.getKey();
+                QueryGroupStats singleQueryGroupStats = new QueryGroupStats(Map.of(queryGroupId, entry.getValue()));
+                allQueryGroups.add(new WlmStats(stat.getNode(), singleQueryGroupStats));
+            }
+        }
+
+        allQueryGroups = sortStats(allQueryGroups, sortBy, sortOrder);
+
+        int startIndex = 0;
+        if (requestedToken != null) {
+            OptionalInt foundIndex = findIndex(allQueryGroups, requestedToken.lastQueryGroupId);
+            if (foundIndex.isEmpty()) {
+                throw new OpenSearchParseException("Invalid or outdated token: " + nextToken);
+            }
+            startIndex = foundIndex.getAsInt();
+        }
+
+        int endIndex = Math.min(startIndex + pageSize, allQueryGroups.size());
+        List<WlmStats> page = allQueryGroups.subList(startIndex, endIndex);
+
+        if (endIndex < allQueryGroups.size()) {
+            String lastQueryGroup = allQueryGroups.get(endIndex - 1).getQueryGroupStats().getStats().keySet().iterator().next();
+            this.responseToken = new PageToken(WlmStrategyToken.generateEncryptedToken(lastQueryGroup), DEFAULT_PAGINATED_ENTITY);
+        } else {
+            this.responseToken = null;
+        }
+
+        return page;
+    }
+
+    /**
+     * Finds the index of a given query group in the list.
+     */
+    private OptionalInt findIndex(List<WlmStats> stats, String queryGroupId) {
+        for (int i = 0; i < stats.size(); i++) {
+            if (stats.get(i).getQueryGroupStats().getStats().containsKey(queryGroupId)) {
+                return OptionalInt.of(i + 1);
+            }
+        }
+        return OptionalInt.empty();
+    }
+
+    @Override
+    public PageToken getResponseToken() {
+        return responseToken;
+    }
+
+    @Override
+    public List<WlmStats> getRequestedEntities() {
+        return paginatedStats;
+    }
+
+    /**
+     * Retrieves paginated stats.
+     */
+    public List<WlmStats> getPaginatedStats() {
+        return paginatedStats;
+    }
+
+    /**
+     * Retrieves the next pagination token.
+     */
+    public PageToken getNextToken() {
+        return responseToken;
+    }
+
+    public static class WlmStrategyToken {
+        private static final String JOIN_DELIMITER = "|";
+        private static final String SPLIT_REGEX = "\\|";
+
+        private final String lastQueryGroupId;
```

Comment:
Should we also consider something else which can help determine the state change and need to start fresh ? For example queryGroupCount